projects/sandbox/train/train/glitch_loader.py

```diff
@@ -0,0 +1,175 @@
+        self,
+        batch_size: int,
+        batches_per_epoch: int,
+        files: Dict[str, List[Path]],
+    ) -> None:
+        super().__init__()
+        self.batch_size = batch_size
+        self.batches_per_epoch = batches_per_epoch
+        self.files = files
+        self.channels = list(files.keys())
+
+        sizes = {}
+        probs = {}
+
+        # get sizes of each file for each channel
+        for channel, fnames in self.files.items():
+            sizes = []
+            for f in fnames:
+                with h5py.File(f, "r") as f:
+                    size = len(f["glitches"])
+                    sizes.append(size)
+
+                    # infer kernel length
+                    self.kernel_length = f["glitches"].shape[-1]
+
+            total = sum(sizes)
+            probs[channel] = np.array([i / total for i in sizes])
+        self.probs = probs
+
+    def __len__(self) -> int:
+        return self.batches_per_epoch
+
+    def sample_fnames(self):
+        fnames = {}
+        for channel, files in self.files.items():
+            fnames[channel] = np.random.choice(
+                files,
+                p=self.probs[channel],
+                size=(self.batch_size,),
+                replace=True,
+            )
+        return fnames
+
+    def sample_batch(self):
+        # allocate memory up front
+        x = np.zeros((self.batch_size, len(self.channels), self.kernel_length))
+
+        # for each channel, sample files for each batch element
+        fnames = self.sample_fnames()
+
+        for i, files in enumerate(fnames.values()):
+            unique_files, inv = np.unique(files, return_inverse=True)
+
+            for j, file in enumerate(unique_files):
+                batch_indices = np.where(inv == j)[0]
+
+                with h5py.File(file, "r") as f:
+                    num = len(f["glitches"])
+                    indices = np.random.randint(num, size=len(batch_indices))
+                    for b, idx in zip(batch_indices, indices):
+                        x[b, i] = f["glitches"][idx]
+        return torch.Tensor(x)
+
+    def __iter__(self):
+        worker_info = torch.utils.data.get_worker_info()
+        if worker_info is None:
+            num_batches = self.batches_per_epoch
+        else:
+            num_batches, remainder = divmod(
+                self.batches_per_epoch, worker_info.num_workers
+            )
+            if worker_info.id < remainder:
+                num_batches += 1
+
+        for _ in range(num_batches):
+            yield self.sample_batch()
+
+
+class ChunkedGlitchSampler(torch.utils.data.IterableDataset):
+    """
+    Wrapper dataset that will loop through chunks of glitches
+    produced from another iterable and randomly sample batches.
+
+    Args:
+        chunk_it:
+            Iterator that will produce chunks of glitches to sample from.
+            Each chunk should have shape
+            `(N, C, T)` where `N` is the number of chunks
```

Comment:
This terminology is a little confusing, with each chunk having N chunks. Maybe the upper category could be called a set? Open to options